drivers/at/at.c

```diff
@@ -0,0 +1,238 @@
+    } while (res > 0);
+}
+
+ssize_t at_send_cmd_get_resp(at_dev_t *dev, const char *command, char *resp_buf, size_t len, uint32_t timeout)
+{
+    ssize_t res = -1;
+
+    at_drain(dev);
+
+    res = at_send_cmd(dev, command, timeout);
+    if (res) {
+        goto out;
+    }
+
+    res = at_readline(dev, resp_buf, len, timeout);
+    if (res == 0) {
+        /* skip possible empty line */
+        res = at_readline(dev, resp_buf, len, timeout);
+    }
+
+out:
+    return res;
+}
+
+ssize_t at_send_cmd_get_lines(at_dev_t *dev, const char *command, char *resp_buf, size_t len, uint32_t timeout)
+{
+    ssize_t res = -1;
+    size_t bytes_left = len - 1;
+    char *pos = resp_buf;
+
+    memset(resp_buf, '\0', len);
+
+    at_drain(dev);
+
+    res = at_send_cmd(dev, command, timeout);
+    if (res) {
+        goto out;
+    }
+
+    while(1) {
+        res = at_readline(dev, pos, bytes_left, timeout);
+        if (res == 0) {
+            continue;
+        }
+        else if (res > 0) {
+            bytes_left -= res;
+            if ((res == 2) && (strncmp(pos, "OK", 2) == 0)) {
+                res = len - bytes_left;
+                break;
+            }
+            else if ((res == 5) && (strncmp(pos, "ERROR", 5) == 0)) {
+                return -1;
+            }
+            else if (strncmp(pos, "+CME ERROR:", 11) == 0) {
+                return -1;
+            }
+            else {
+                pos += res;
+                if (bytes_left) {
+                    *pos++ = '\n';
+                    bytes_left--;
+                }
+                else {
+                    return -1;
+                }
+            }
+        }
+        else {
+            break;
+        }
+    }
+
+out:
+    return res;
+}
+
+int at_send_cmd_wait_prompt(at_dev_t *dev, const char *command, uint32_t timeout)
+{
+    unsigned cmdlen = strlen(command);
+
+    at_drain(dev);
+
+    uart_write(dev->uart, (const uint8_t *)command, cmdlen);
+    uart_write(dev->uart, (const uint8_t *)AT_END_OF_LINE, sizeof(AT_END_OF_LINE) - 1);
+
+    if (at_expect_bytes(dev, command, cmdlen, timeout)) {
+        return -1;
+    }
+
+    if (at_expect_bytes(dev, AT_END_OF_LINE "\n", sizeof(AT_END_OF_LINE), timeout)) {
+        return -2;
+    }
+
+    if (at_expect_bytes(dev, ">", 1, timeout)) {
+        return -3;
+    }
+
+    return 0;
+}
+
+int at_send_cmd_wait_ok(at_dev_t *dev, const char *command, uint32_t timeout)
+{
+    int res;
+    char resp_buf[32];
+
+    res = at_send_cmd_get_resp(dev, command, resp_buf, sizeof(resp_buf), timeout);
+    if (res > 0) {
+        if (strcmp(resp_buf, "OK") == 0) {
+            res = 0;
+        }
+        else {
+            res = -1;
+        }
+    }
+
+    return res;
+}
+
+ssize_t at_readline(at_dev_t *dev, char *resp_buf, size_t len, uint32_t timeout)
+{
+    ssize_t res = -1;
+    char *resp_pos = resp_buf;
+
+    memset(resp_buf, 0, len);
+
+    while (len) {
```

Comment:
What happens if the response string is longer than 'len' and the while loop is ended. If I understand this code well enough then I'm afraid there is no NUL byte terminating the string.

Reply:
That's what the memset in line 210 should take care of?

Reply:
No, not quite. Say there is a 'len' of 5 and on the input comes "abcdefg\r\n". The while loop ends as soon as "abcde" was read and resp_buf is completely filled up (no NUL byte anymore). No?

Reply:
Ah, right. I guess ```-ENOSPC``` would be the desired result in that case?

Reply:
Looking at the code again, if no full line could be read, the result of the function is -1. Shouldn't that be enough? Anyhow, I've added a condition setting the first byte of resp_buf to the null byte in that case.

Reply:
I still don't see how the string is terminated with a nul-byte. In my example above resp_buf contains the 5 characters a, b, c, d and e. The return value is 5. Maybe you say that the caller is responsible to not treat the result as a nul-terminated string?

Reply:
The return variable ```res```, which defaults to ```-1```, is only set to the number of bytes read if a full line could be read. In your example (```[ a, b, c, d, e ]```), the while loop would end with ```res == -1``` and thus set ```*resp_buf = '\0'```.

Reply:
@keestux Does this address your concerns?

Reply:
I can see that there can be no unterminated string, even when a longer line comes in. Good.

On the other hand, you take drastic measures to throw away the whole input (instead of truncating).

If this is what we want then OK.
BTW. The doc says "<0 on error", which is either a timeout (-ETIMEDOUT) or a too long input line (-1). Perhaps you want to put this in the function description (at.h).